readme.md
# Adaptive Images

Is a solution to automatically create, cache, and deliver device-appropriate versions of your website's content images. It does not require you to change your mark-up. It is intended for use with [Responsive Designs](http://www.abookapart.com/products/responsive-web-design) and to be combined with [Fluid Image](http://unstoppablerobotninja.com/entry/fluid-images/) techniques.

## Benefits

* Ensures you and your visitors are not wasting bandwidth delivering images at a higher resolution than the vistor needs.
* Will work on your existing site, as it requires no changes to your mark-up.
* Is device agnostic (it works by detecting the size of the visitors screen)
* Is CMS agnostic (it manages its own image re-sizing, and will work on any CMS or even on flat HTML pages)
* Is entirely automatic. Once added to your site, you need do no further work.
* Highly configurable
** Set the resolutions you want to become adaptive (usually the same as the ones in your CSS @media queries)
** Choose where you want the cached files to be stored
** Configure directories to ignore (protect certain directories so AI is not applied to images within them)

Find out more, and view examples at [http://adaptive-images.com](http://adaptive-images.com)

## Legal

Adaptive Images by Matt Wilcox is licensed under a "Creative Commons Attribution 3.0 Unported License":http://creativecommons.org/licenses/by/3.0/

##Instructions

### Basic instructions

Copy `adaptive-images.php` and `.htaccess` into the root directory of your site. If you already have a htaccess file DO NOT OVERWRITE IT, skip down to the advanced instructions.

Copy the following Javascript into the `<head>` of your site. It MUST go in the head as the first bit of JS, before any other JS. This is because it needs to work as soon as possible, any delay wil have adverse effects.

    <script>document.cookie='resolution='+Math.max(screen.width,screen.height)+'; path=/';</script>

That's it, you're done. You should proberbly configure some preferences though.

NOTE 1: if you would like to take advantage of high-pixel density devices such as the iPhone4 or iPad3 Retina display you can use the following JavaScript instead.

It will send higher-resolution images to such devices - be aware this will mean slower downloads for Retina users, but better images.

    <script>document.cookie='resolution='+Math.max(screen.width,screen.height)+("devicePixelRatio" in window ? ","+devicePixelRatio : ",1")+'; path=/';</script>

NOTE 2: you do not need the `ai-cookie.php` file unless you are using the alternate method of detecting the users screen size. So delete it if you like, no one likes mess.

NOTE 3: If you are extra paranoid about security you can have the ai-cache directory sit outside of your web-root so it's not publicly accessible. Just set the paths properly in the .htaccess file and the .php file.

### Additional settings and configuration

### .htaccess
Instructions are in the file as comments (any line that starts with a # is a comment, and doesn't actually do anything)
Follow the instructions inside that code to specify any directories you don't want to use Adaptive-Images on.

### PHP
You can now open the php file and have a play with the settings that are in the CONFIG area. By default it looks like this:

`$resolutions` are the screen widths we'll work with. In the default it will store a re-sized image for large screens, normal screens, tablets, phones, and tiny phones.
In general, if you're using a responsive design in CSS, these breakpoints will be exactly the same as the ones you use in your media queries.

`$cache_path` If you don't like the cached images being written to that folder, you can put it somewhere else. Just put the path to the folder here and make sure you create it on the server if for some reason that couldn't be done autmoatically by adaptive-images.php.

`$sharpen` Will perform a subtle sharpening on the rescaled images. Usually this is fine, but you may want to turn it off if your server is very very busy.

`$watch_cache` If your server gets very busy it may help performance to turn this to FALSE. It will mean however that you will have to manually clear out the cache directory if you change a resource file


### Troubleshooting

Most of the time people report problems it is due to one of two things:

**If images vanish**

There is something wrong with your .htaccess configuration. This happens mostly on WordPress sites - it's because the server, and wordpress, have specific requirements that are different from most servers. You'll have to play about in the .htaccess file and read up on how to use ModRewrite.

**If you're seeing error images (big black ones)**

That's AI working, so your .htaccess is fine. Read the messages on the image. Most of the time you'll only see this problem because your server requires less strict permissions to write to the disc. Try setting the ai-cache directory to 775, and if all else fails use 777 - but be aware this is not very secure if you're on a shared server, and you ought to instead contact your administrator to get the server set up properly.


changelog.txt
version 1.0.2 (2013-04-20 ------------------------------------------------------------

	* Fixed fatal fallback problem: fallback now actually works
	* Introduced fallback default resolutions for mobile and desktop
	  Can be overridden by size terms if needed
	* Better handling of crop ratios:
	  Ratio gets normalized, this means fraction is reduced to something like x:1
	  Maximum allowed digits of the resulting ratio is 1, means something like 1.6:1
	  Resulting benefit: file structure is cleaner; ratios can be "used again"
	  

version 1.0.1 (2013-04-19) -----------------------------------------------------------

	* Introduced ['scalings']: now the breakpoint sizes don’t dictate the resulting
	  image widths any more. ['scalings'] are a corresponding array to ['resolutions']
	  containing the desired image widths.
	* Introducing "fullsize request" via query string.
	  Reserved query params for this are: 'original', 'full', 'fullsize', 'source' and 'src'.
	  Don’t use them as size terms, they are hardcoded and overridden it anyway. 
	  Example usage: image.jpg?original or image.jpg?size=original etc.
	* Added "Fancy Wish To Do List".
	
	# Appendum # to v1.0
	* By default in setup.php: generation of ['breakpoints'] from 1% to 100%
	  with smart sharpening according to resize amount.
	

version 1.0 (2013-04-19) --------------------------------------------------------------

	* Refactored setup.php: new comments, new instructions
	* Refactored .htaccess to make it more feasible
	* Updated readme.textile, updated changelog.txt to officially manifest the fork

---------------------------------------------------------------------------------------




//////////////////////////////////////////////////////////////
//                                                         //
//  Adaptive Images (über-extended) fork starts here      //
//                                                       //
//////////////////////////////////////////////////////////






version 1.5.2.1 (2012/04/18) ------------------------------------------------------------

  NEW FEATURE
    * Now you can use parameters on the image-src to recalculate the resolution of the 
      image on different break-points.
      You can use % depending on the resolution or fixe width in pixel.
      
      Setup the break-points in the CONFIG section ofadaptive-images.php
      $breakpoints  = array('default' => 0, 'small' => 840, 'large' => 960);

      Use the break-points names and set a width in your html-template.
      <img src="image.jpg?bp=default-100%_small-480px_large-960px" />

    * Debug Mode
      CONFIG $debug_mode = TRUE;
      Writes image dimensions into the cached images.

    * Prevent Cache for Development
      CONFIG $prevent_cache = TRUE;
      AI will recalculate and deliver new images on each call for development.

version 1.5.2 (2012/04/16) ------------------------------------------------------------

  BUG FIX
    * AI was not generating high DPI images larger than the existing maximum
      $resolutions value
      
      I used a < when I needed a >. That is now fixed. If AI detects retina displays it
      is now able to generate larger images than the top bounds set in $resolutions.

version 1.5.1 (2012/04/14) ------------------------------------------------------------

  ENHANCEMENT
    * JPGs are now created as Progressive rather than Optimised.
      
      Progressive JPGs are actually slightly smaller in file-size than Optimised JPGs,
      and have the benefit of appearing to load faster in modern browsers (they
      download in passes, gradually getting sharper).

version 1.5 (2012/04/14) --------------------------------------------------------------

  ENHANCEMENT
    * Much improved support for Retina and other high DPI devices

      NOTE: Retina support requires using alternate JavaScript:
      <script>document.cookie='resolution='+Math.max(screen.width,screen.height)+("devicePixelRatio" in window ? ","+devicePixelRatio : ",1")+'; path=/';</script>

      NOTE: If you see folders in the ai-cache directory which are *larger* than any
      defined in your $resolutions array, it is because someone with a high DPI / Retina
      display visited your site, and AI was able to generate the required image for them.

      DETAILS:
      Previous AI behaviour meant that supporting Retina required adding very high values
      in the $resolutions array. This is sub-optimal because:

      1) This array is meant to be nothing more than the same values as your media query
         breakpoints. Adding larger ones isn't intuitive.
      2) When no cookie is set, AI sends non-mobile devices the largest $resolution value,
         which in that case could be far too large for most people (2000px+)

      AI is now much smarter, you do not need to edit the $resolutions array; just leave
      that as your CSS MQ sizes.

      AI now auto-detects the pixel density of any high DPI device and either picks a
      matching size from the existing array, or creates new high-dpi images which are
      multiples of your chosen $resolutions values.

version 1.4.1 (2012/03/28) ------------------------------------------------------------

  ENHANCEMENT
    * More helpful error messages if something goes wrong.
    * Updated the documentation a little.

version 1.4 (2012/02/19) --------------------------------------------------------------

  ENHANCEMENT
    * Auto-creates the cache folder with the correct permissions.
      No need to manually create the ai-cache folder any more, AI does it for you.

  BUG FIX
    * Fixed an incorrect permission on created files, they're now 755 not 777.

version 1.3.9 (2012/02/17) ------------------------------------------------------------
                                                                         Thanks to Gerv

  ENHANCEMENT
    * Better detection of mobile device in the event cookies are not available.
      

adaptive-images-uber-extended/adaptive-images/setup.php
<?php
 
/*    ##    R E A D M E   N O T E S   A N D   H O W   T O   S E T U P    ##
 *
 *
 * *Breakpoints and sizes*
 * Now you can define, what size an image has to have on specific configurable
 * design breakpoints equal to your CSS media queries. Your can give them
 * custom names (the "size term") and set the their widths in pixel or percentage.
 * If you use for instance 100% as a default case, the image will be 100% of the
 * size of the resolution, that is equal or closest higher than the present users
 * device maximum width – but maximum width of the largest configured resolution
 * breakpoint but also not larger than the original full-res image size.
 * 
 * 
 * *Cropping (cutting to fit)*
 * You can define an aspect ratio for a term. All depending images will be cropped
 * to that ratio after resizing is done.
 * 
 * 
 * *Sharpen*
 * You can define if you want to sharpen the resized images and even the amount of
 * sharpness applied. 0 is no sharpening at all, up to 30 is subtile, more than 30 is
 * really impacting, max is 500 (imagine that!). Usually doing this is fine,
 * but you may want to turn it off if your server is very very busy.
 * 
 * *JPEG quality*
 * You can define the JPEG quality for a size term when device pixel ratio is at 1.
 * If above 1 (= "retina"), you can set a different JPEG compression level – very useful
 * for applying "Netvlies Retina Revolution Trick"
 * (http://blog.netvlies.nl/design-interactie/retina-revolution/) if you want to use it.
 * JPEG quality 100: of course is best; down to 70: is pleasant;
 * below 40: really compressed (and small in size).
 * 
 * 
 * *Prevent cache*
 * For developing, you might want to disable caching, to see changes of the
 * setup and the images on  * every page reload. For this, setup the following
 * in the Main Settings section:
 * 
 * $config['browser_cache'] = 0;
 * $config['prevent_cache'] = TRUE;
 * 
 * But remember to turn caching on again, when the site is going live!
 * 
 * 
 * *Debug mode*
 * This inserts some information into the generated images like: image dimensions,
 * the aspect ratio, the device-width and pixel density.
 * 
 * Turn this off when site goes live!
 * 
 * 
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */





	// MAIN SETTINGS
	// where to store the generated image files; specify from your DOCUMENT_ROOT!
	// Make sure the folders have at least 755 permission!
	// If you don't like the cached images being written to that folder, you can put it somewhere else.
	// Just put the path to the folder here and make sure you create it on the server
	// if for some reason that couldn't be done automatically by adaptive-images.php.
	$config['cache_path']           = 'adaptive-images/ai-cache';
	
	// checks if the adapted image in cache aren’t outdated; ensures updated source images are cached again.
	// In case your server gets very busy, it may help performance to turn this to FALSE.
	// It will mean however, that you will have to manually clear out the cache directory if you change a resource file	
	$config['watch_cache']          = TRUE;			// default: true
	
	// other caching settings
	$config['browser_cache']        = 60 * 60 * 24;	// period of time in seconds the images will stay in cache of browsers	
	//$config['browser_cache']		= 0;			// enable this line line during developement!
	$config['prevent_cache']        = FALSE; 		// default: false; true: images will resized on every image request

	// while developing: inserts information like image dimensions, ratio and the device-width into the image
	$config['debug_mode']           = FALSE;		// default: false




 
	// ORIGINAL BEHAVIOR – Matts original device resolution depending solution
	// true: to have this original automatic resizer enabled; this is the default
	// false: to only use the size terms in a query string and serve the original full res pic at a "native request"
	$config['enable_resolutions']   = TRUE;  

	// Here are our breakpoints for the default behavior; screen widths in pixels; ascending order
	$config['resolutions']          = array(0, 320, 480, 640, 1080, 1440, 2048, 2880);

	// Now new in AIue: kind of 'one size fits all' values, corresponding to the resolutions above
	// These image widths are served, when a specific resolution is present
	$config['scalings']             = array(0, 320, 480, 640,  960, 1440, 1920, 2880); 

	// This is part of Johanns extended version introducing the "size terms"
	// Configure the internal breakpoint names here
	// These values should be equal to the resolutions above, to not confuse things more than necessary
	// NOTE: the corresponting ['scalings'] from above do apply anyway!
	// That’s why the array sizes MUST be equal in any case!
	$config['breakpoints'] = array(
	                                 'default' =>    0,
	                                 'micro'   =>  320,
	                                 'mini'    =>  480,
	                                 'small'   =>  640,
	                                 'medium'  => 1080,
	                                 'normal'  => 1440,
	                                 'large'   => 2048,
	                                 'huge'    => 2880
	                              );
							
	// Check if the array sizes are equal, because this is very important!
	// If you receive the error message, align the sizes of the three arrays above!
	$line = __LINE__; 	$line = $line - 1;
	$error_str = "Error in Adaptive Images: all array sizes MUST be equal to work correctly!<br>
			    Check " . __FILE__ . " at line " . $line . " to solve this.";	 
	if ( count($config['breakpoints']) != count($config['scalings']) or (count($config['resolutions']) != count($config['scalings'])) ) { exit($error_str); }
	
	
	// Setting the two fallback widths for mobile and desktop
	// can be overridden by size terms later
	// MUST be pixel or unitless (meaning pixel)
	$config['fallback']['mobile']  =  '480px';
	$config['fallback']['desktop'] = '1440px';
	
							
	// Some settings concerning the default behavior image quality
	$config['jpg_quality']          = 90;			// quality of a generated JPG at device pixel ratio of 1; values: 0 to 100; default: 80
	$config['jpg_quality_retina']   = 30;			// use for netvlies' compression trick; 100 to 0; default: 50
	$config['sharpen']['status']    = TRUE;			// enables sharpening of resized images
	$config['sharpen']['amount']    = 20;			// 0 is none, 30 is pleasant, max is 500
	
	
	
	
	
	// EXTENDED BEHAVIOR
	// Define the size terms: these are the query strings used to append on the image file
	// NOTE: 'original', 'full', 'fullsize', 'source' and 'src' are reserved terms!
	// They’re hardcoded to serve the original source image in any case!
	// We need that, while having $config['enable_resolutions'] = TRUE; Matts original behavior
	// Below this is just an example how to set it up
	// You MUST use units like % or px! Otherwise it gets ignored and results in '100%'
	// Usage afterwards: <img src="image.jpg?size=term" />
	$setup['term']['breakpoints']['default'] = '100%';
	$setup['term']['breakpoints']['normal'] = '1024px';
	$setup['term']['ratio'] = '1.6181:1';
	$setup['term']['jpg_quality'] = 95;
	$setup['term']['jpg_quality_retina'] = 40;
	$setup['term']['sharpen']['amount'] = 40;
	$setup['term']['fallback']['mobile'] = '320px';		// must be pixel
	$setup['term']['fallback']['desktop'] = '1024px';	// must be pixel

	// set up your own size terms here; use $config['breakpoints'] = array( '{strings}' ) as breakpoints
	
	


	// UBER-EXTENDED BEHAVIOR
	// Prepare percentage breakpoints as a built-in default to have it present
	// Maybe useful for fluid layouts
	// A smart amout of sharpening is applied, according to the amoutn of resizing done
	// NOTE: use carfully! This percentage things spams your ai-cache directory with all kinds of calculated resolutions!
	for ($percent = 1; $percent < 100; $percent++) {
		$setup[$percent . '%']['breakpoints']['default'] = $percent . '%';
		$setup[$percent . '%']['sharpen']['amount'] = floor( (100 - $percent) / 1.3); // slightly sharpen according to size    
		}
	
	
	
	
?>